### PR TITLE
[Merged by Bors] - Add Shared Reference Encoding

### PR DESCRIPTION
**Pull request by pascal-huber**
changes:
 - Added `OpenFracRef` and `CloseFracRef` statements
 - Added `FracRef` predicate
 - Added `frac_bor_atomic_acc` method
 - Shared reference encoding
 - Fix mutable references encoding

### CODE REVIEW

File: prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs [hunk 1]
```diff
@@ -894,7 +919,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> BuiltinMethodsInterface for Lowerer<'p, 'v, 'tcx> {
                 vir_mid::TypeDecl::TypeVar(_) => {
                     // move_place of a generic variable has no body
                 }
-                vir_mid::TypeDecl::Tuple(decl) => {
+                vir_mid::TypeDecl::Tuple(ref decl) => {
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs]: Instead of adding `ref` to every branch, you could change `match type_decl` to `match &type_decl`.
```suggestion
                vir_mid::TypeDecl::Tuple(decl) => {
```

File: prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs [hunk 2]
```diff
@@ -867,23 +884,31 @@ impl<'p, 'v: 'p, 'tcx: 'v> BuiltinMethodsInterface for Lowerer<'p, 'v, 'tcx> {
...
+            if let vir_mid::TypeDecl::Reference(_) = type_decl {
+                statements.push(stmtp! { position =>
+                    unfold OwnedNonAliased<ty>(source_place, source_root_address, source_value, lifetime)
+                });
+            } else {
+                statements.push(stmtp! { position =>
+                    unfold OwnedNonAliased<ty>(source_place, source_root_address, source_value)
+                });
+            }
             match type_decl {
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs]:
```suggestion
            match &type_decl {
```

File: prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs [hunk 3]
```diff
@@ -468,6 +483,76 @@ impl IntoLow for vir_mid::Statement {
                     Ok(vec![])
                 }
             }
+            Self::OpenFracRef(statement) => {
+                let place = statement.place.get_parent_ref().unwrap();
+                let ty = place.get_type();
+                lowerer.encode_frac_bor_atomic_acc_method(ty)?;
+                lowerer.encode_newlft_method()?; // ???
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs]: Why is this needed?

File: prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs [hunk 4]
```diff
@@ -468,6 +483,76 @@ impl IntoLow for vir_mid::Statement {
...
+                let lifetime = lowerer.encode_lifetime_const_into_variable(statement.lifetime)?;
+                let perm_amount = vir_low::Expression::fractional_permission(statement.rd_perm);
+                let reference_place = lowerer.encode_expression_as_place(place)?;
+                let reference_value = place.to_procedure_snapshot(lowerer)?;
+                let tmp_var_name = format!("tmp_frac_ref_perm${}", lifetime.name);
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs:494]: Is this guaranteed to be unique?

File: prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs [hunk 5]
```diff
@@ -468,6 +483,76 @@ impl IntoLow for vir_mid::Statement {
...
+                )?;
+                let address_snapshot =
+                    lowerer.reference_address_snapshot(ty, reference_value, statement.position)?;
+                let tmp_var_name = format!("tmp_frac_ref_perm${}", lifetime.name);
+                let tmp_frac_ref_perm =
+                    lowerer.create_variable(tmp_var_name, vir_low::Type::Perm)?;
+                let type_decl = lowerer.encoder.get_type_decl_mid(ty)?;
+                let target_type = &type_decl.unwrap_reference().target_type;
+                Ok(vec![
+                    // TODO: use macros if possible
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs]: https://github.com/viperproject/prusti-dev/blob/ef018b531fb9cb7233a51f1d210a55d303d50370/prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs#L254-L269

Comment (vakaras) [reply prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs]: The variant of `acc` with permission amount is defined [here](https://github.com/viperproject/prusti-dev/blob/ef018b531fb9cb7233a51f1d210a55d303d50370/vir/src/low/macros.rs#L76-L82)

File: prusti-viper/src/encoder/middle/core_proof/types/interface.rs [hunk 6]
```diff
@@ -172,7 +172,9 @@ impl<'p, 'v: 'p, 'tcx: 'v> Private for Lowerer<'p, 'v, 'tcx> {
                 } else {
                     let parameters = vars! {
                         address: Address,
-                        target_current: {target_type.clone()}
+                        target_current: {target_type.clone()},
+                        // NOTE: final destructor is also necessary for shared references
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/types/interface.rs]: Why is this the case?

File: vir/defs/high/operations_internal/identifier/ty.rs [hunk 7]
```diff
@@ -129,7 +129,8 @@ impl WithIdentifier for ty::Slice {
 impl WithIdentifier for ty::Reference {
     fn get_identifier(&self) -> String {
         format!(
-            "ref${}${}",
+            "ref${}${}${}",
+            self.uniqueness,
             self.lifetime.get_identifier(),
```
Comment (vakaras) [inline vir/defs/high/operations_internal/identifier/ty.rs]: The lifetime probably should not be included in the reference's identifier.

**Review by vakaras — CHANGES_REQUESTED** [review]
Some minor comments, otherwise look good.

File: prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs (hunk 3 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs]: It is not -> removed.

File: prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs (hunk 1 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs]: done

File: prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs (hunk 2 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs]: done

File: prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs (hunk 4 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs:494]: I believe so - at least I cannot come up with a counterexample. The name must be the same in `OpenFracRef` and `CloseFracRef` - thats why I didn't use `create_new_temporary_variable`.

File: vir/defs/high/operations_internal/identifier/ty.rs (hunk 7 above)
Comment (pascal-huber) [reply vir/defs/high/operations_internal/identifier/ty.rs]: yes, I removed it. done

File: prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs (hunk 4 above)
Comment (vakaras) [reply prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs:494]: Ok.

File: prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs (hunk 5 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/middle/core_proof/into_low/cfg.rs]: I added a macro which supports both typed predicate names and fractional permission.

File: prusti-viper/src/encoder/middle/core_proof/types/interface.rs (hunk 6 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/middle/core_proof/types/interface.rs]: Assignments did not distinguish between shared and mutable references. Should be ok now.

This is for shared references:

```
  // bb0[4] _2 = &'_#3r _1
  tmp$0 := assign$ref$Shared$I32$ref_$I32(_2$place(), _2$address, _1$place(), _1$address, _1$snapshot$1, lft_3$snapshot$0, write / 3)
  inhale _2$snapshot$1 == tmp$0
  inhale _1$snapshot$2 == destructor$Snap$ref$Shared$I32$$target_current(tmp$0)
```

**Review by vakaras — APPROVED** [review]
LGTM
bors r+
